### DISCUSSION TITLE
How to have multi-monitor workspaces(groups)  like dwm

### DISCUSSION THREAD

**Discussion by aahsnr-configs** [qtile/qtile#4352] [Q&A] [answered]
Hi,

I have a triple monitor setup, and I was wondering if it's possible to have groups behave like how workspaces behave in dwm? Basically the following is a summary of what I am looking for:
1. For each monitor, i.e., each screen, I want to label 6 groups as 123456.
2. I want these 6 groups to stick to their respective monitors
3. And when I am focused on a specific monitor, I want to change groups using mod4+label without affecting the groups in the other 2 monitors.

**Comment by elParaguayo** [comment]
Definitely doable. I'll post some code later.

**Comment by elParaguayo** [comment, marked as answer]
OK - this is a very basic proof-of-concept:

```python
NUM_SCREENS = 3
NUM_GROUPS = 8
GROUP_MAP = {}
GROUP_LIST = []

# Here we create two objects:
# 1) A dict, "GROUP_MAP", which has a key relating to the screen index. The value is another dict
#    where groups are numbered from 1 to NUM_GROUPS. The value is a unique group name.
# 2) A list of all groups. Each group has a unique name but the label is between 1 and NUM_GROUPS
for screen in range(NUM_SCREENS):
    screen_dict = {}
    for group in range(NUM_GROUPS):
        name = f"{screen * NUM_GROUPS + group}"
        group_num = group + 1
        screen_dict[group_num] = name
        GROUP_LIST.append(Group(name, label=str(group_num)))
    GROUP_MAP[screen] = screen_dict

@lazy.function
def switch_group(qtile, group_num):
    """
    This function looks up the unique group name for group number on the current screen
    and then displays that group.
    """
    group_name = GROUP_MAP[qtile.current_screen.index + 1][group_num]
    qtile.groups_map[group_name].toscreen()


@lazy.window.function
def move_window_to_group(window, group_num, switch_group=False, toggle=False):
    """
    This function looks up the unique group name for group number on the current screen
    and then moves the window to that group.
    """   
    group_name = GROUP_MAP[window.qtile.current_screen.index + 1][group_num]
    window.togroup(group_name, switch_group, toggle)

# Groups should be set to our group list
groups = GROUP_LIST

# We bind keys mod + 1-8 to call our function to display a group depending on the focused screen
# mod + shift + 1-8 moves window to that group on the active screen
for i in range(NUM_GROUPS):
    keys.append(Key([mod], str(i + 1), switch_group(i + 1)))
    keys.append(Key([mod, "shift"], str(i + 1), move_window_to_group(i + 1, switch_group=True)))

# mod + control + 1-3 moves window to current group on that screen
for i in range(NUM_SCREENS):
    keys.append(Key([mod, "control"], str(i + 1), lazy.window.toscreen(i)))
```

For the `GroupBox` widget, you'll need to set `visible_groups` to `GROUP_MAP[screen_index].values()` for each screen where `screen_index` starts at 0.

**Reply by aahsnr-configs** [reply]
would you mind clarify the part about screen_index? Is the screen_index referring to the start of the group number at 0, so each screen has an index that starts at 0. Or are you referring to the screen index my 3 monitors, where monitor 1 starts 0 and so, on
@elParaguayo

**Comment by aahsnr-configs** [comment]
I am still learning python. Will this be able to solve the issue of moving windows between screens that you raised in reddit?

**Reply by elParaguayo** [reply]
No. That's not included there.  That will depend on how you want to move windows between screens and groups.

**Reply by aahsnr-configs** [reply]
But why do you think it is tricky?

**Reply by elParaguayo** [reply]
It could be very easy.  Just depends on what you want.  If you want a different keybinding to move a window to every possible group, that could be hard as you'll run out of keys. Moving to current group on a particular screen is easy.

**Reply by aahsnr-configs** [reply]
`Moving to current group on a particular screen is easy.` 

This is what I am looking for. On an unrelated topic, since I have a triple monitor setup with an ultrawide in the middle and 16:9 on the sides, so I am planning to one layout for the ultrawide and master/stack on the 16:9 monitors. I was wondering if layouts can only be assigned to groups instead of screens? If that's correct I am planning to write python code to do that so that I can brush my python skills. How tough do u think that might be?

Since I have your attention how stable is qtile as a compositor? I have read some older posts complaining and others saying only good things. As for the animations, does the compositor version really have any like sway?

BTW, I wanted to thank you for being so helpful. I haven't had support like this where the maintainer of the project writes code for a total stranger in his busy schedule.

**Reply by elParaguayo** [reply]
> `Moving to current group on a particular screen is easy.`
> 
> This is what I am looking for. 

I've edited the code above to do this. It's all untested so there may be a bug or two in there.

> On an unrelated topic, since I have a triple monitor setup with an ultrawide in the middle and 16:9 on the sides, so I am planning to one layout for the ultrawide and master/stack on the 16:9 monitors. I was wondering if layouts can only be assigned to groups instead of screens? If that's correct I am planning to write python code to do that so that I can brush my python skills. How tough do u think that might be?

Layouts are generally global but groups can limit which ones are used in that group. So, you could limit certain layouts for those groups on a particular screen.

> Since I have your attention how stable is qtile as a compositor? I have read some older posts complaining and others saying only good things. As for the animations, does the compositor version really have any like sway?

So, I assume you're using Wayland here... I'm not best to comment on it as I don't use it. However, I think it's pretty stable and there is a lot of work being done on it. I don't think there are any animations... yet (@m-col would be better placed to comment).

> BTW, I wanted to thank you for being so helpful. I haven't had support like this where the maintainer of the project writes code for a total stranger in his busy schedule.

You're very welcome.

**Reply by aahsnr-configs** [reply]
so basically the additional code would theoretically allow me to move windows between screens?

**Reply by elParaguayo** [reply]
Yes, ctrl + mod + 1-3 should move the window to each of those screens.

**Reply by aahsnr-configs** [reply]
I am going to try your code in a couple of days. Is it okey if I close the discussion after I have tested them?

**Reply by elParaguayo** [reply]
No need to close.  Just let me know when you've tried it.

**Comment by aahsnr-configs** [comment]
Btw are then any plans to add animations to qtile wayland?
